Add AudioVideoRendererRemote

jyavenard opened the pull request:
#### 6069cb8422b29100d2c6d22e20029d5eef06e496
<pre>
Add AudioVideoRendererRemote
<a href="https://bugs.webkit.org/show_bug.cgi?id=298907">https://bugs.webkit.org/show_bug.cgi?id=298907</a>
<a href="https://rdar.apple.com/160651588">rdar://160651588</a>

Reviewed by Youenn Fablet.

And adopt it with MediaPlayerPrivateWebM.
Code is behind a preference, disabled by default.

With preference enabled, the MediaPlayerPrivateWebM will run in the web content
process including all media parsing. Only the rendering is occurring in the GPU process.

On visionOS, docking the video is for now non-functional.

Will be covered by existing tests, once preference is enabled.

Canonical link: <a href="https://commits.webkit.org/300810@main">https://commits.webkit.org/300810@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/751ffcac0ee65f8cb2c2bcc8ac920c98f2ce2501

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123939 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43654 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34351 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130730 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/76091 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/125816 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44378 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52248 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/94277 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/76091 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/b43abddd-e2d0-4ce8-a360-d051959af1a6) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126893 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35360 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/110872 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74878 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/9133f34b-1677-48d6-ba13-f5cfb57da619) 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/123295 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/34309 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29034 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74202 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/116098 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/105090 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29256 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133422 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/122471 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50892 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38761 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/102741 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51266 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107092 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102565 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26084 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47912 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26164 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/47745 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50745 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56509 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/153567 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50219 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39035 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53565 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51893 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

jernoble commented on Source/WebCore/platform/MediaStrategy.cpp:
```diff
@@ -83,4 +87,30 @@ void MediaStrategy::addMockMediaSourceEngine()
+bool MediaStrategy::hasRemoteRendererFor(MediaPlayerMediaEngineIdentifier type) const
+{
+    return m_remoteRenderersEnabled & 1 << static_cast<uint32_t>(type);
+}
+
+void MediaStrategy::enableRemoteRenderer(MediaPlayerMediaEngineIdentifier type, bool enabled)
+{
+    if (enabled)
+        m_remoteRenderersEnabled |= 1 << static_cast<uint32_t>(type);
+    else
+        m_remoteRenderersEnabled &= ~(1 << static_cast<uint32_t>(type));
+}
+#endif
```
This seems like a good use case for WTF/BitSet.h.

jyavenard replied:
I do in the latest version, I applied all the comments you mentioned during our last meeting.

jernoble commented on Source/WebKit/GPUProcess/media/RemoteAudioVideoRendererProxyManager.cpp:
```diff
@@ -0,0 +1,576 @@
+            return;
+        }
+        renderer->notifyWhenHasAvailableVideoFrame({ });
+    }
```
I would prefer early returns over deeply nested if statements. Something like:

```suggestion
    RefPtr renderer = rendererFor(identifier);
    if (!renderer)
        return;
    if (!notify) {
        renderer->notifyWhenHasAvailableVideoFrame({ });
        return;
    }

    renderer->notifyWhenHasAvailableVideoFrame([weakThis = WeakPtr { *this }, identifier](auto presentationTime, auto clockTime) {
        if (RefPtr protectedThis = weakThis.get(); protectedThis && protectedThis->m_renderers.contains(identifier))
            protectedThis->m_gpuConnectionToWebProcess.get()->connection().send(Messages::AudioVideoRendererRemoteMessageReceiver::HasAvailableVideoFrame(presentationTime, clockTime, protectedThis->stateFor(identifier)), identifier);
    });
```

jyavenard replied:
I re-wrote them all to always exit early, except if it's a one-liner

youennf commented on Source/WebCore/platform/MediaStrategy.cpp:
```diff
@@ -83,4 +87,27 @@ void MediaStrategy::addMockMediaSourceEngine()
 #endif
 
+#if ENABLE(VIDEO)
+RefPtr<AudioVideoRenderer> MediaStrategy::createAudioVideoRenderer(WTF::LoggerHelper* loggerHelper) const
```
Why not LoggerHelp&?

jyavenard replied:
so it can be null for the windows build that don't have logging enabled.
I didn't want to have to constructor, one with release_log enabled taking two arguments, and one without release_log

youennf commented on Source/WebKit/GPUProcess/media/RemoteAudioVideoRendererProxyManager.h:
```diff
@@ -0,0 +1,181 @@
+namespace WebKit {
+
+class GPUConnectionToWebProcess;
+class LayerHostingContext;
```
Needed, we have LayerHostingContextManager.h ?

youennf commented on Source/WebKit/GPUProcess/media/RemoteAudioVideoRendererProxyManager.h:
```diff
@@ -0,0 +1,181 @@
+    void didReceiveMessage(IPC::Connection&, IPC::Decoder&) final;
+    void didReceiveSyncMessage(IPC::Connection&, IPC::Decoder&, UniqueRef<IPC::Encoder>&) final;
+
+    // Messages
```
Do we need all those message methods to be public, it seems they could all be private?

jyavenard replied:
yes they can be.

youennf commented on Source/WebKit/WebProcess/GPU/media/AudioVideoRendererRemote.cpp:
```diff
@@ -0,0 +1,839 @@
+    return adoptRef(*new AudioVideoRendererRemote(loggerHelper, connection, identifier));
+}
+
+AudioVideoRendererRemote::AudioVideoRendererRemote(LoggerHelper* loggerHelper, GPUProcessConnection& connection, RemoteAudioVideoRendererIdentifier identifier)
```
LoggerHelper&?

jyavenard replied:
ditto

youennf commented on Source/WebKit/WebProcess/GPU/media/AudioVideoRendererRemote.cpp:
```diff
@@ -0,0 +1,839 @@
+
+void AudioVideoRendererRemote::notifyFirstFrameAvailable(Function<void()>&& callback)
+{
+    m_firstFrameAvailableCallback = WTFMove(callback);
```
ASSERT(! m_firstFrameAvailableCallback)?

jyavenard replied:
it's not incorrect to call it multiple times, with different callback, in particular it can be called with `{ }` which disable the callback, so I don't want to assert there

youennf commented on Source/WebKit/WebProcess/GPU/media/AudioVideoRendererRemote.cpp:
```diff
@@ -0,0 +1,839 @@
+
+void AudioVideoRendererRemote::notifyWhenHasAvailableVideoFrame(Function<void(const MediaTime&, double)>&& callback)
+{
+    m_hasAvailableVideoFrameCallback = WTFMove(callback);
```
ASSERT(! m_hasAvailableVideoFrameCallback)?

Ditto for other callbacks?

jyavenard replied:
same explanation as above. A call with nullptr/{ } turn off the callback, this is used constantly by the API as it kind of mimics the AVSampleBufferDisplayLayer one.

youennf commented on Source/WebKit/WebProcess/GPU/media/AudioVideoRendererRemote.cpp:
```diff
@@ -0,0 +1,839 @@
+}
+
+#if ENABLE(VIDEO_PRESENTATION_MODE)
+void AudioVideoRendererRemote::setVideoFullscreenLayer(PlatformLayer* videoFullscreenLayer, WTF::Function<void()>&& completionHandler)
```
Could completionHandler be a CompletionHandler?

jyavenard replied:
that callback comes from the HTMLMediaElement `WEBCORE_EXPORT void setVideoFullscreenLayer(PlatformLayer*, Function<void()>&& completionHandler = [] { });` it probably should be

youennf commented on Source/WebKit/WebProcess/GPU/media/AudioVideoRendererRemote.cpp:
```diff
@@ -0,0 +1,839 @@
+        return;
+    gpuProcessConnection->connection().send(Messages::RemoteAudioVideoRendererProxyManager::EnqueueSample(m_identifier, trackIdentifier, MediaSamplesBlock::fromMediaSample(sample), expectedMinimum), 0);
+    if (auto it = m_requestMediaDataWhenReadyData.find(trackIdentifier); it != m_requestMediaDataWhenReadyData.end())
+        it->value.pendingSamples++;
```
Is there a possibility that we enqueue before filling `m_requestMediaDataWhenReadyData`? What is the impact?

jyavenard replied:
it is okay to fill data before calling the callback.
It tracks how the AVSBDL API work:
```
void enqueueSample() {
    while ([AVSBDL isReadyForMoreData])
        [AVSBDL enqueueSample:...]
    [AVSBDL requestMediaDataWhenReady:[] {
        [AVSBDL stopRequestingMediaData];
        enqueueSample();
    }];
}
```

jyavenard replied:
calling any of those callback with a nullptr is equivalent of cancelling the callback.

youennf commented on Source/WebKit/WebProcess/GPU/media/AudioVideoRendererRemote.cpp:
```diff
@@ -0,0 +1,839 @@
+void AudioVideoRendererRemote::setVideoFullscreenFrame(const FloatRect& frame)
+{
+    RefPtr gpuProcessConnection = m_gpuProcessConnection.get();
+    if (!isGPURunning() || !gpuProcessConnection)
```
What is the strategy when GPU process crashes?

jyavenard replied:
Good point. there are none. I will treat it as a fatal error, the site can then reload the video.

jernoble commented on Source/WebKit/GPUProcess/media/RemoteAudioVideoRendererProxyManager.cpp:
```diff
@@ -0,0 +1,576 @@
+        completionHandler(WTFMove(result));
+        return;
+    }
+    completionHandler(std::nullopt);
```
This is a little more edge case, but I'd still refactor this to be an early-return in most cases:

```suggestion
    RefPtr renderer = rendererFor(identifier);
    if (!renderer) {
        completionHandler(std::nullopt);
        return;
    }

   
    if (RefPtr videoFrame = renderer->currentVideoFrame()) {
        completionHandler(m_videoFrameObjectHeap->add(videoFrame.releaseNonNull()));
        return;
    }

    completionHandler(std::nullopt);    
```

jernoble commented on Source/WebKit/GPUProcess/media/RemoteAudioVideoRendererProxyManager.cpp:
```diff
@@ -0,0 +1,576 @@
+
+    bool canShowWhileLocked = false;
+    // FIXME: Provide canShowWhileLocked, from external configuration.
+    // #if PLATFORM(IOS_FAMILY)
+    //         m_configuration.canShowWhileLocked;
+    // #else
+    //         false;
+    // #endif
```
Probably can skip checking this in.

jyavenard replied:
this is addressed in a follow-up PR [#51493](https://github.com/WebKit/WebKit/pull/51493)

jernoble commented on Source/WebKit/GPUProcess/media/RemoteAudioVideoRendererProxyManager.h:
```diff
@@ -0,0 +1,181 @@
+#include "MessageReceiver.h"
+#include "RemoteAudioVideoRendererIdentifier.h"
+#include "RemoteAudioVideoRendererState.h"
+#include "RemoteVideoFrameProxy.h"
```
This doesn't look to be used.

jyavenard replied:
it is for `std::optional<RemoteVideoFrameProxy::Properties>` later

jernoble commented on Source/WebKit/GPUProcess/media/RemoteAudioVideoRendererProxyManager.h:
```diff
@@ -0,0 +1,181 @@
+#include <WebCore/MediaPromiseTypes.h>
+#include <wtf/Forward.h>
+#include <wtf/Logger.h>
+#include <wtf/MediaTime.h>
```
And this just needs a forward declaration.

jernoble commented on Source/WebKit/GPUProcess/media/RemoteAudioVideoRendererProxyManager.h:
```diff
@@ -0,0 +1,181 @@
+#include <wtf/Forward.h>
+#include <wtf/Logger.h>
+#include <wtf/MediaTime.h>
+#include <wtf/ObjectIdentifier.h>
```
And this is already included by `RemoteAudioVideoRendererIdentifier.h`.

jernoble commented on Source/WebKit/GPUProcess/media/RemoteAudioVideoRendererProxyManager.h:
```diff
@@ -0,0 +1,181 @@
+    RemoteAudioVideoRendererProxyManager(GPUConnectionToWebProcess&);
+    ~RemoteAudioVideoRendererProxyManager();
+
+    using TrackIdentifier = WebCore::AudioVideoRenderer::TrackIdentifier;
```
Pulling in `AudioVideoTrackRenderer.h` just to get access to an inner type is a good sign that type needs to be pulled out of `AudioVideoTrackRenderer.h` and made into something that can be forward-declared or pulled in from a cheaper header.

jernoble commented on Source/WebKit/GPUProcess/media/RemoteAudioVideoRendererProxyManager.h:
```diff
@@ -0,0 +1,181 @@
+    std::optional<SharedPreferencesForWebProcess> sharedPreferencesForWebProcess() const;
+
+private:
+    struct RendererContext {
+        RefPtr<WebCore::AudioVideoRenderer> renderer;
+#if PLATFORM(COCOA)
+        LayerHostingContextManager layerHostingContextManager;
+#endif
+    };
```
`LayerHostingContextManager` should be a `std::unique_ptr<LayerHostingContextManager>` so that it can use a forward declaration, rather than pulling in all of `LayerHostingContextManager.h` for this private structure.

jernoble commented on Source/WebKit/GPUProcess/media/RemoteAudioVideoRendererProxyManager.h:
```diff
@@ -0,0 +1,181 @@
+
+#if ENABLE(GPU_PROCESS) && ENABLE(VIDEO)
+
+#include "Connection.h"
```
This can be a forward-declaration instead.

jernoble commented on Source/WebKit/GPUProcess/media/RemoteAudioVideoRendererProxyManager.cpp:
```diff
@@ -0,0 +1,576 @@
+            if (!protectedThis->m_renderers.contains(identifier))
+                return;
+            protectedThis->m_gpuConnectionToWebProcess.get()->connection().send(Messages::AudioVideoRendererRemoteMessageReceiver::ErrorOccurred(error), identifier);
+        }
```
```suggestion
        if (RefPtr protectedThis = weakThis.get(); protectedThis && protectedThis->m_renderers.contains(identifier))
            protectedThis->m_gpuConnectionToWebProcess.get()->connection().send(Messages::AudioVideoRendererRemoteMessageReceiver::ErrorOccurred(error), identifier);
```

jernoble commented on Source/WebKit/GPUProcess/media/RemoteAudioVideoRendererProxyManager.cpp:
```diff
@@ -0,0 +1,576 @@
+#if PLATFORM(COCOA)
+            protectedThis->contextFor(identifier).layerHostingContextManager.setVideoLayerSizeIfPossible();
+#endif
+            protectedThis->m_gpuConnectionToWebProcess.get()->connection().send(Messages::AudioVideoRendererRemoteMessageReceiver::FirstFrameAvailable(protectedThis->stateFor(identifier)), identifier);
```
```suggestion
        if (RefPtr protectedThis = weakThis.get(); protectedThis && protectedThis->m_renderers.contains(identifier)) {
#if PLATFORM(COCOA)
            protectedThis->contextFor(identifier).layerHostingContextManager.setVideoLayerSizeIfPossible();
#endif
            protectedThis->m_gpuConnectionToWebProcess.get()->connection().send(Messages::AudioVideoRendererRemoteMessageReceiver::FirstFrameAvailable(protectedThis->stateFor(identifier)), identifier);
```

jernoble replied:
Ditto the above to a number of the functions below. I might even suggest adding a helper method like:

```cpp
    bool has(RemoteAudioVideoRendererIdentifier identifier) const { return m_renderers.contains(identifier); }
```

...to make this pattern a little less verbose.

jernoble commented on Source/WebKit/GPUProcess/media/RemoteAudioVideoRendererProxyManager.cpp:
```diff
@@ -0,0 +1,576 @@
+
+void RemoteAudioVideoRendererProxyManager::requestMediaDataWhenReady(RemoteAudioVideoRendererIdentifier identifier, TrackIdentifier trackIdentifier)
+{
+    if (RefPtr renderer = rendererFor(identifier)) {
```
Early return.

jernoble commented on Source/WebKit/GPUProcess/media/RemoteAudioVideoRendererProxyManager.cpp:
```diff
@@ -0,0 +1,576 @@
+            }
+        });
+        return;
+    }
```
I'd move the early return out of the deeply nested `if()` and simplify.

jernoble commented on Source/WebKit/GPUProcess/media/RemoteAudioVideoRendererProxyManager.cpp:
```diff
@@ -0,0 +1,576 @@
+    if (RefPtr renderer = rendererFor(identifier)) {
+        renderer->performTaskAtTime(time, [completionHandler = WTFMove(completionHandler)](auto& time) mutable {
+            completionHandler(time);
+        });
```
Why can't we pass the `completionHandler` directly into `performTaskAtTime()`?

jernoble commented on Source/WebKit/GPUProcess/media/RemoteAudioVideoRendererProxyManager.cpp:
```diff
@@ -0,0 +1,557 @@
+    };
+
+    context.renderer->notifyWhenErrorOccurs([weakThis = WeakPtr { *this }, identifier](PlatformMediaError error) {
+        if (RefPtr protectedThis = weakThis.get(); protectedThis && protectedThis->m_renderers.contains(identifier))
```
This pattern of `protectedThis->m_renderers.contains(identifier)` happens a lot. Maybe you could add a convenience method like:

```cpp
private:
    bool has(RemoteAudioVideoRendererIdentifier identifier) const { return m_renderers.contains(identifier); }
```

That would turn this call site (and others) into:


```suggestion
        if (RefPtr protectedThis = weakThis.get(); protectedThis && protectedThis->has(identifier))
```

jernoble commented on Source/WebKit/GPUProcess/media/RemoteAudioVideoRendererProxyManager.cpp:
```diff
@@ -0,0 +1,557 @@
+        renderer->cancelTimeReachedAction();
+}
+
+void RemoteAudioVideoRendererProxyManager::performTaskAtTime(RemoteAudioVideoRendererIdentifier identifier, const MediaTime& time)
+{
+    RefPtr renderer = rendererFor(identifier);
+    if (!renderer)
+        return;
+    renderer->performTaskAtTime(time, [weakThis = WeakPtr { *this }, time, identifier](auto) {
+        if (RefPtr protectedThis = weakThis.get(); protectedThis && protectedThis->m_renderers.contains(identifier))
+            protectedThis->m_gpuConnectionToWebProcess.get()->connection().send(Messages::AudioVideoRendererRemoteMessageReceiver::TaskTimeReached(time, protectedThis->stateFor(identifier)), identifier);
+    });
+}
```
This seems pretty weird? Why isn't this passing in a callback like `seekTo(...)`?

jernoble commented on Source/WebKit/GPUProcess/media/RemoteAudioVideoRendererProxyManager.h:
```diff
@@ -0,0 +1,182 @@
+    void setPreferences(RemoteAudioVideoRendererIdentifier, WebCore::VideoMediaSampleRendererPreferences);
+    void setHasProtectedVideoContent(RemoteAudioVideoRendererIdentifier, bool);
+
+    // TracksRendererInterface
+    using TrackIdentifier = WebCore::AudioVideoRenderer::TrackIdentifier;
```
Pulling in `AudioVideoRenderer.h` just to get access to this type seems to indicate that type needs to be pulled out of the `AudioVideoRenderer` class and put in a cheaper header.

jernoble commented on Source/WebKit/GPUProcess/media/RemoteLegacyCDMProxy.cpp:
```diff
@@ -88,10 +88,11 @@ RefPtr<MediaPlayer> RemoteLegacyCDMProxy::cdmMediaPlayer(const LegacyCDM*) const
 {
-    if (!m_factory)
+    RefPtr factory = m_factory.get();
+    if (!factory)
         return std::nullopt;
 
-    return m_factory->sharedPreferencesForWebProcess();
+    return factory->sharedPreferencesForWebProcess();
```
This seems wrong. Why did `factory` lose its `m_`?

jyavenard replied:
I use `factory` which is the RefPtr for it ; it was giving me a static analyser failure, which is what I changed it to `RefPtr factory = m_factory.get()`

jernoble commented on Source/WebKit/WebProcess/GPU/media/AudioVideoRendererRemote.h:
```diff
@@ -0,0 +1,251 @@
+#if !RELEASE_LOG_DISABLED
+    , private LoggerHelper
+#endif
+    , public GPUProcessConnection::Client
```
`GPUProcessConnection::Client` should be pulled out of that class and into a new `GPUProcessConnectionClient.h` header, and `GPUProcessConnection` can be forward-declared.

jyavenard replied:
I'll take this as a follow-up, this seems out of scope.

jernoble commented on Source/WebKit/WebProcess/GPU/media/AudioVideoRendererRemote.h:
```diff
@@ -0,0 +1,251 @@
+#if ENABLE(GPU_PROCESS) && ENABLE(VIDEO)
+
+#include "GPUProcessConnection.h"
+#include "LayerHostingContext.h"
```
This doesn't seem to be needed.

jyavenard replied:
it is for `WebCore::HostingContext m_layerHostingContext;` member.

jernoble commented on Source/WebKit/WebProcess/GPU/media/AudioVideoRendererRemote.h:
```diff
@@ -0,0 +1,251 @@
+
+class AudioVideoRendererRemote final
+    : public WebCore::AudioVideoRenderer
+    , public VideoLayerRemoteParent
```
This class also needs to be pulled out into its own header, rather than pulling in all of `VideoLayerRemote.h`.

jernoble commented on Source/WebKit/WebProcess/GPU/media/AudioVideoRendererRemote.h:
```diff
@@ -0,0 +1,251 @@
+
+    WTF_ABSTRACT_THREAD_SAFE_REF_COUNTED_AND_CAN_MAKE_WEAK_PTR_IMPL;
+
+    class MessageReceiver final : public IPC::WorkQueueMessageReceiver<WTF::DestructionThread::Any> {
```
This is defined backwards. This class should be defined in `AudioVideoRendererRemoteMessageReceiver.h` and pulled into this class with a `using` statement, rather than vice-versa. That would avoid pulling in `WorkQueueMessageReceiver.h` into this header.

jyavenard replied:
I wanted them to be in the same file, it makes for much easier code readability so it's all there at a glance.

jyavenard replied:
Note that this is unused yet, but the plan is to make AudioVideoRendererRemote thread-safe so that it can be used with MSE in a worker.

Having said that `WorkQueueMessageReceiver` is a pretty small header with very little depth.

jyavenard committed:
Add AudioVideoRendererRemote
https://bugs.webkit.org/show_bug.cgi?id=298907
rdar://160651588

Reviewed by Youenn Fablet.

And adopt it with MediaPlayerPrivateWebM.
Code is behind a preference, disabled by default.

With preference enabled, the MediaPlayerPrivateWebM will run in the web content
process including all media parsing. Only the rendering is occurring in the GPU process.

On visionOS, docking the video is for now non-functional.

Will be covered by existing tests, once preference is enabled.

Canonical link: https://commits.webkit.org/300810@main
